File: android/app/src/main/java/com/lockoutprotocol/guardian/push/Pusher.kt
package com.lockoutprotocol.guardian.push

import com.lockoutprotocol.guardian.data.Prefs
import java.util.concurrent.TimeUnit
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody

/**
 * Zero-setup push alerts via ntfy (https://ntfy.sh). No account, no password — the user just
 * subscribes the free ntfy app to their private topic code. We POST the alert text to
 * {server}/{topic}. The topic string is unguessable, so it doubles as the secret.
 */
object Pusher {

    private val client = OkHttpClient.Builder()
        .callTimeout(20, TimeUnit.SECONDS)
        .build()

    fun send(prefs: Prefs, title: String, message: String): Result<Unit> = runCatching {
        require(prefs.pushEnabled) { "push disabled" }
        val topic = prefs.ntfyTopic
        require(topic.isNotBlank()) { "no topic" }

        val url = prefs.ntfyServer.trimEnd('/') + "/" + topic
        val req = Request.Builder()
            .url(url)
            .addHeader("Title", title.toAsciiHeader())
            .addHeader("Priority", "high")
            .addHeader("Tags", "rotating_light")
            .post(message.toRequestBody("text/plain; charset=utf-8".toMediaType()))
            .build()

        client.newCall(req).execute().use { resp ->
            if (!resp.isSuccessful) error("ntfy ${resp.code}: ${resp.body?.string()}")
        }
    }

    /** ntfy header values must be ASCII; strip anything else so titles never break the request. */
    private fun String.toAsciiHeader(): String =
        filter { it.code in 32..126 }.ifBlank { "Guardian" }
}

File: ios/Lockout/Push/Pusher.swift
import Foundation

/// Push alerts via ntfy: no account, no email server, your partner subscribes a free app to a
/// private topic.
///
/// No dependency on `Prefs` or any actor, so the shield action extension can send from its own
/// process with the app not running.
enum Pusher {

    struct Config: Sendable {
        let enabled: Bool
        let server: String
        let topic: String
    }

    /// For the extensions, which can't reach the app's `Prefs`.
    static func configFromShared() -> Config {
        let defaults = UserDefaults(suiteName: SessionStore.appGroup) ?? .standard
        let enabled = defaults.object(forKey: "push_enabled") == nil
            ? true : defaults.bool(forKey: "push_enabled")
        return Config(enabled: enabled,
                      server: defaults.string(forKey: "ntfy_server") ?? "https://ntfy.sh",
                      topic: Keychain.get("ntfy_topic") ?? "")
    }

    /// Returns nil on success, or a reason. Never throws.
    @discardableResult
    static func send(_ cfg: Config, title: String, message: String) async -> String? {
        guard cfg.enabled else { return "alerts are switched off" }
        guard !cfg.topic.isEmpty else { return "no alert code is set" }
        guard let url = URL(string: cfg.server.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
                            + "/" + cfg.topic) else {
            return "the ntfy server URL isn't valid"
        }

        var req = URLRequest(url: url, timeoutInterval: 20)
        req.httpMethod = "POST"
        req.httpBody = Data(message.utf8)
        // ntfy header values must be ASCII; one em dash would otherwise break the request.
        req.setValue(asciiHeader(title), forHTTPHeaderField: "Title")
        req.setValue("high", forHTTPHeaderField: "Priority")
        req.setValue("rotating_light", forHTTPHeaderField: "Tags")

        do {
            let (_, response) = try await URLSession.shared.data(for: req)
            let code = (response as? HTTPURLResponse)?.statusCode ?? 0
            return (200...299).contains(code) ? nil : "ntfy returned HTTP \(code)"
        } catch {
            return "couldn't reach \(cfg.server)"
        }
    }

    private static func asciiHeader(_ s: String) -> String {
        let filtered = s.unicodeScalars.filter { $0.value >= 32 && $0.value <= 126 }
        let out = String(String.UnicodeScalarView(filtered))
        return out.isEmpty ? "Lockout" : out
    }
}
